Add `system.map(...)` for transforming the output of a system

JoJoJet opened the pull request:
# Objective

Any time we wish to transform the output of a system, we currently use system piping to do so:

```rust
my_system.pipe(|In(x)| do_something(x))
```

Unfortunately, system piping is not a zero cost abstraction. Each call to `.pipe` requires allocating two extra access sets: one for the second system and one for the combined accesses of both systems. This also adds extra work to each call to `update_archetype_component_access`, which stacks as one adds multiple layers of system piping.

## Solution

Add the `AdapterSystem` abstraction: similar to `CombinatorSystem`, this allows you to implement a trait to generically control how a system is run and how its inputs and outputs are processed. Unlike `CombinatorSystem`, this does not have any overhead when computing world accesses which makes it ideal for simple operations such as inverting or ignoring the output of a system.

Add the extension method `.map(...)`: this is similar to `.pipe(...)`, only it accepts a closure as an argument instead of an `In<T>` system.

```rust
my_system.map(do_something)
```

This has the added benefit of making system names less messy: a system that ignores its output will just be called `my_system`, instead of `Pipe(my_system, ignore)`

---

## Changelog

TODO

## Migration Guide

The `system_adapter` functions have been deprecated: use `.map` instead, which is a lightweight alternative to `.pipe`.

```rust
// Before:
my_system.pipe(system_adapter::ignore)
my_system.pipe(system_adapter::unwrap)
my_system.pipe(system_adapter::new(T::from))

// After:
my_system.map(std::mem::drop)
my_system.map(Result::unwrap)
my_system.map(T::from)

// Before:
my_system.pipe(system_adapter::info)
my_system.pipe(system_adapter::dbg)
my_system.pipe(system_adapter::warn)
my_system.pipe(system_adapter::error)

// After:
my_system.map(bevy_utils::info)
my_system.map(bevy_utils::dbg)
my_system.map(bevy_utils::warn)
my_system.map(bevy_utils::error)
```

JoJoJet commented on crates/bevy_utils/src/synccell.rs:
```diff
@@ -7,6 +7,7 @@
 /// Provides a wrapper that allows making any type unconditionally [`Sync`] by only providing mutable access.
 #[repr(transparent)]
+#[derive(Clone, Copy)]
 pub struct SyncCell<T: ?Sized> {
```
I'd like extra attention paid here to make sure that it's sound to clone a `SyncCell`. I don't see why this would be unsound, but I also can't figure out how to prove that it *is* sound.

alice-i-cecile commented on crates/bevy_ecs/src/schedule/condition.rs:
```diff
@@ -940,97 +937,29 @@ pub mod common_conditions {
+
+#[doc(hidden)]
+#[derive(Clone, Copy)]
+pub struct NotMarker;
```
Doc strings here, even though it's hidden will help future contributors.

alice-i-cecile commented on crates/bevy_ecs/src/system/adapter_system.rs:
```diff
@@ -0,0 +1,124 @@
+use super::{ReadOnlySystem, System};
+
+pub trait Adapt<S: System>: Send + Sync + 'static {
```
Doc strings here: linking out to the map method would be most useful.

alice-i-cecile commented on crates/bevy_ecs/src/system/adapter_system.rs:
```diff
@@ -0,0 +1,124 @@
+}
+
+#[derive(Clone)]
+pub struct AdapterSystem<Func, S> {
```
```suggestion
/// A system that takes the output of `S` and transforms it by applying `Func` to it.
pub struct AdapterSystem<Func, S> {
```

alice-i-cecile commented on crates/bevy_ecs/src/system/mod.rs:
````diff
@@ -286,6 +311,7 @@ pub mod adapter {
     ///     println!("{x:?}");
     /// }
     /// ```
+    #[deprecated = "use `.map(...)` instead"]
````
Clever: I like being able to deprecate these.

cart commented on crates/bevy_ecs/src/system/adapter_system.rs:
```diff
@@ -0,0 +1,126 @@
+        self.system.name()
+    }
+
+    fn type_id(&self) -> std::any::TypeId {
```
This should probably be `TypeId::of::<Self>`, as this is _not_ a system of type `S`. Using the internal `System::name` also gives me pause, but that feels way less "potentially behavior breaking". And generating a unique name would allocate, which I know would undermine one of your goals for this abstraction.

JoJoJet replied:
The bit about names is a fair objection. I think that losing the `!` for `NotSystem` could be problematic when reporting errors, since it may not be obvious that a run condition is inverted.

I'll try to think on an API for adapting system names.

cart reviewed (approved):
Seems like a good complement to `pipe`. Simpler interface (avoids `In<T>`). `.map` is a common rust convention. Saves on some expenses.

I think in a world where we find a way to remove `In<T>` (in favor of just `T`) I might advocate for removing `map`, as I'm not sure avoiding the startup allocations for `name()` is worth "duplicate" abstractions. We could probably make `name` lazily computed if that is a concern as most games won't even need system names at runtime.

alice-i-cecile reviewed (changes requested):
I agree with Cart on the type_id comment, then this LGTM.

JoJoJet commented:
I've fixed the `type_id` method, and I added an API that lets system adapters change the system name. Now, calling `not(resource_exists<T>)` will correctly result in a system called `!resource_exists<T>`.  
Using `.map()` will *not* change the name of the inner system -- IMO this is cleaner, but I'm willing to change it if you guys think it's important for the system name to include the mapping function.

Also, I'm going to temporarily mark this as a draft. I want to see if I can make this abstraction usable for #8705 -- this will require giving the system adapter access to the system's name, but I don't have time to experiment with that at the moment.

JoJoJet commented:
Alright, I've simplified the API for system names which will make it easier to use this in the aforementioned PR (and makes it consistent with `CombinatorSystem`.

I believe I have addressed all of the concerns raised :).

ItsDoot commented:
What's the status on this PR? Interesting in using `.map`.

TimJentzsch commented:
Is it intentional that the changelog is still "TODO"?

JoJoJet commented:
I don't think we still do changelogs. Not sure why it's still in the PR template.

cart commented:
We do still do changelogs, we just lean on auto-generation more. Including changelogs in PRs is still welcome and encouraged as it will make the auto-generated changelog better.
